### PR TITLE
Search Tree Performance - FolderMatches and FileMatches

### LINKED ISSUE #162246: Search Tree Performance - FolderMatches
When testing the performance for the search tree, I noticed that some of the new search tree model functions take up a lot of time:

image

Which shows that the calls to `foldermatches()` takes up a lot of time. Should investigate to reduce search performance.

### PR DESCRIPTION
**Pull request by andreamah**
Fixes #162246

Changing the calls to `folderMatches()` and `fileMatches()` to return an iterator to reduce performance overhead of search tree changes.

### CODE REVIEW

File: src/vs/workbench/contrib/search/common/searchModel.ts [hunk 1]
```diff
@@ -603,15 +604,25 @@ export class FolderMatch extends Disposable {
 	}
 
 	matches(): (FileMatch | FolderMatchWithResource)[] {
-		return [...this.fileMatches(), ...this.folderMatches()];
+		return [...this.fileMatchesIterator(), ...this.folderMatchesIterator()];
+	}
+
+	fileMatchesArray(): FileMatch[] {
+		// for testing only; try to use fileMatchesIterator for performance reasons
```
Comment (roblourens) [inline src/vs/workbench/contrib/search/common/searchModel.ts:611]: Comments like this are for the person calling the function, so use jsdoc

Comment (roblourens) [reply src/vs/workbench/contrib/search/common/searchModel.ts:611]: You can also create the array in the test code itself from calling the iterator version, I don't see why this has to be in product code

File: src/vs/workbench/contrib/search/common/searchModel.ts [hunk 2]
```diff
@@ -770,7 +792,7 @@ export class FolderMatch extends Disposable {
 
 		const removed = [];
 		for (const match of fileMatches as FileMatch[]) {
-			if (this.fileMatches().includes(match)) {
+			if (this.getFileMatch(match.resource)) {
```
Comment (roblourens) [inline src/vs/workbench/contrib/search/common/searchModel.ts:795]: Can't this just access `this._fileMatches` directly? If not, looks like getFileMatch should be private at least

File: src/vs/workbench/contrib/search/common/searchModel.ts (hunk 1 above)
Comment (andreamah) [reply src/vs/workbench/contrib/search/common/searchModel.ts:611]: there are some tests that get the foldermatches of foldermatches and so on, so it looks cleaner if it's something that can be called directly on the object. I can move it over, though!

**Review by roblourens — DISMISSED** [review]

**Comment by roblourens** [comment]
Looking at the changes, it makes sense that those methods were being called more often to build the full tree model, so makes sense that this is a good perf improvement!

**Review by roblourens — APPROVED** [review]

**Pull request merged**